lib/spack/spack/build_environment.py

```diff
@@ -525,22 +525,29 @@ def child_fun():
     carries on.
     """
 
-    def child_execution(child_connection):
+    def child_execution(child_connection, input_stream):
         try:
             setup_package(pkg, dirty=dirty)
-            function()
+            function(input_stream)
             child_connection.send([None, None, None])
         except Exception as e:
             child_connection.send([type(e), e, None])
         finally:
             child_connection.close()
 
     parent_connection, child_connection = multiprocessing.Pipe()
-    p = multiprocessing.Process(
-        target=child_execution,
-        args=(child_connection,)
-    )
-    p.start()
+    try:
+        # Forward sys.stdin to be able to activate / deactivate
+        # verbosity pressing a key at run-time
+        input_stream = os.fdopen(os.dup(sys.stdin.fileno()))
```

Comment:
**TL;DR** It's needed because of `multiprocessing` and the file descriptor closed in the parent should be a duplicate of `sys.stdin`, not `sys.stdin` itself.

-------

Disclaimer: I am not that confident with low level POSIX APIs, so I may be missing more idiomatic methods, or misinterpreting some of the docs.

> Why do we need to forward stdin to the child to get keypresses working?

That's because `multiprocessing.Process` does:
```python
sys.stdin.close()
sys.stdin = open(os.devnull)
```
during `_bootstrap()`, see [docs](https://docs.python.org/2/library/multiprocessing.html#all-platforms) here (search for `sys.stdin`). I didn't know that until @adamjstewart pointed to the bug for pressing <kbd>v</kbd> and I read all the docs for `multiprocessing` when I noticed that stdin was not being modified by `termios`. 

The least invasive solution that came to my mind was to "duplicate" `sys.stdin` file descriptor in the parent and forward it to the child. As far as I understood it's the duplicate file descriptor that I am closing, not `sys.stdin` as a call to `os.dup` should be the same as a call to `dup`, which in turn is a shorthand for:
```c
fcntl(fildes, F_DUPFD, 0);
```
see [the docs for dup](http://pubs.opengroup.org/onlinepubs/9699919799/functions/dup.html) and [the docs for fcntl](http://pubs.opengroup.org/onlinepubs/009695399/functions/fcntl.html).